ingest/src/main/java/com/microsoft/azure/kusto/ingest/IngestionProperties.java

```diff
@@ -301,40 +308,50 @@ void validate() throws IngestionClientException {
-        String message = null;
+        String message = "";
 
-        if (ingestionMapping.getColumnMappings() != null) {
-            if (StringUtils.isNotBlank(mappingReference)) {
-                message += String.format("Both mapping reference '%s' and column mappings were defined.", mappingReference);
-            }
-            for (ColumnMapping column : ingestionMapping.getColumnMappings()) {
-                if (!column.isValid(ingestionMappingKind)) {
-                    message += String.format("Column mapping '%s' is invalid", column.columnName);
-                }
-            }
-        } else if (StringUtils.isBlank(mappingReference)) {
+        if ((ingestionMapping.getColumnMappings() == null) && StringUtils.isBlank(mappingReference)) {
             if (dataFormat.isMappingRequired()) {
                 message += String.format("Mapping must be specified for '%s' format.", dataFormat.name());
             }
 
-            if (ingestionMappingKind != null && ingestionMappingKind != IngestionMapping.IngestionMappingKind.unknown) {
-                message += String.format("If IngestionMappingKind was defined ('%s'), mapping must be specified.", ingestionMappingKind);
+            if (ingestionMappingKind != null) {
+                message += String.format("IngestionMappingKind was defined ('%s'), so a mapping must be defined as well.", ingestionMappingKind);
             }
-        }
+        } else { // a mapping was provided
+            if (dataFormat.getIngestionMappingKind() != null && !dataFormat.getIngestionMappingKind().equals(ingestionMappingKind)) {
+                message += String.format("Wrong ingestion mapping for format '%s'; mapping kind should be '%s', but was '%s'.",
+                        dataFormat.name(), dataFormat.getIngestionMappingKind(), ingestionMappingKind != null ? ingestionMappingKind.name() : "null");
+            }
+
+            if (ingestionMapping.getColumnMappings() != null) {
+                if (StringUtils.isNotBlank(mappingReference)) {
+                    message += String.format("Both mapping reference '%s' and column mappings were defined.", mappingReference);
+                }
 
-        if (dataFormat.isMappingRequired() && !dataFormat.getIngestionMappingKind().equals(ingestionMappingKind)) {
-            message += String.format("Wrong ingestion mapping for format '%s'; found '%s' mapping kind.", dataFormat.name(), ingestionMappingKind != null ? ingestionMappingKind.name() : "null");
+                if (ingestionMappingKind == null) {
+                    message += "Column mappings were provided, so the IngestionMappingKind must be provided as well.";
+                } else {
+                    for (ColumnMapping column : ingestionMapping.getColumnMappings()) {
+                        if (!column.isValid(ingestionMappingKind)) {
+                            message += String.format("Column mapping '%s' is invalid", column.columnName);
```

Comment:
Make message a StringBuilder (if you go to the += here idea will do it for you)

Reply:
I have a different approach to some lint smells, and this is one of them.

My raitonale is:
1) Here, it's the generation of an error message, which is not the happy flow and rare.
2) Further, it's limited to just the number of columns. The performance benefit of StringBuilder is negligible in this case, in which case I very much prefer the readability of `+=`.

Reply:
I don't think += is more readable than append.
Especially if you're going to add new lines, you can use TextStringBuilder and `appendLn`